src/librustdoc/html/render/print_item.rs

```diff
@@ -1300,8 +1355,16 @@ fn render_struct(
                     WhereClause { gens: g, indent: 0, end_newline: true }.print(cx.cache())
                 )
             }
-            let mut has_visible_fields = false;
             w.write_str(" {");
+            let count_fields = fields
+                .iter()
+                .filter(|f| matches!(clean::StructFieldItem(..) = *f.kind))
+                .count();
+            let has_visible_fields = count_fields > 0;
+            let toggle = should_hide_fields(count_fields);
+            if toggle {
+                toggle_open(w, "fields");
```

Comment:
Looking at the code, I'd feel much better if we handled it using a closure like it was suggested. It seems very easy to forget... :-/

Reply:
I don't think it's possible to have a closure for more complex things like the trait. I could use a drop guard so that it ICEs instead, but that's about it.

Reply:
What's preventing to have something like:

```rust
fn wrap_with_toggle<T, E, F: FnOnce() -> Result<T, E>>(f: F) -> Result<T, E> {
   // toggle_open code
  let ret = f()?;
  // toggle_close code
  ret
}
```

? I might miss something obvious here. :-/

Reply:
Look at the trait code; it calls `toggle_open` at three different points, and `toggle_close` once. 

Again, I _tried_ this, it was too ugly and even more confusing, so I decided not to. I don't think this is too bad. We could add a destructor bomb but honestly we're using this a couple times and shouldn't have to yet. I'd rather get the entire toggle stuff revamped first.